### PR TITLE
Use numpy tape converter in interfaces instead of tape.Unwrap

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
`tape.Unwrap` calls the data setter of an operator once per data value it has. `tape.set_parameters` can be generically useful, but not for unwrapping tapes to numpy. For example, this is too costly in large `Sum` operators. The new `convert_to_numpy_parameters` is much faster at doing the same thing.

**Description of the Change:**
Replace all uses of `tape.Unwrap` with `convert_to_numpy_parameters(tape)` everywhere

**Benefits:**
Code runs much faster particularly with arithmetic ops

**Possible Drawbacks:**
N/A

Notes:
- ~Depends on #3983~ <-- it's merged 😄 
- further speedups will come from #3912 
- We can deprecate `Unwrap` should we decide to do so

### CODE REVIEW

**Comment by timmysilv** [comment]
[sc-36324]

File: pennylane/interfaces/jax_jit_tuple.py [hunk 1]
```diff
@@ -42,11 +43,11 @@ def _validate_jax_version():
         raise InterfaceUnsupportedError(msg)
 
 
-def _copy_tape(t, a):
+def _copy_tape(t, a, unwrap=True):
     """Copy a given tape with operations and set parameters"""
     tc = t.copy(copy_operations=True)
     tc.set_parameters(a, trainable_only=False)
```
Comment (albi3ro) [inline pennylane/interfaces/jax_jit_tuple.py:49]: Is this just going to have the exact same performance problems we are trying to solve?

File: pennylane/interfaces/jax.py [hunk 2]
```diff
@@ -566,10 +551,10 @@ def _compute_jvps(jacs, tangents, multi_measurements):
     return jvps
 
 
-def _copy_tape(t, a):
+def _copy_tape(t, a, unwrap=True):
```
Comment (albi3ro) [inline pennylane/interfaces/jax.py:554]: Can we rename this helper method to `set_parameters_on_copy_and_unwrap`? And perhaps move it to the top of the file, as code above relies on it. Yes, this helper copies the tape, but it entire point is to set parameters and unwrap to numpy.

File: pennylane/interfaces/jax_jit_tuple.py (hunk 1 above)
Comment (timmysilv) [reply pennylane/interfaces/jax_jit_tuple.py:49]: well, it used to do this `set_parameters` and then _another one_ when calling `tape.Unwrap` on this copied tape, so it's better. But yes, we should try to remove all uses of `tape.set_parameters` (assuming it's used with the ultimate goal of unwrapping). TF has the same problem, but perhaps I should do there what jax is doing, and work on improving this in a subsequent PR.

File: pennylane/interfaces/jax.py (hunk 2 above)
Comment (timmysilv) [reply pennylane/interfaces/jax.py:554]: sure. I've certainly augmented its purpose in this PR, a re-naming would fit too

Comment (albi3ro) [reply pennylane/interfaces/jax.py:554]: This comment repeats bellow.

File: pennylane/interfaces/torch.py [hunk 3]
```diff
@@ -186,8 +187,8 @@ def backward(ctx, *dy):
                 #
                 # so we cannot support higher-order derivatives.
 
-                with qml.tape.Unwrap(*ctx.tapes):
-                    jacs = ctx.gradient_fn(ctx.tapes, **ctx.gradient_kwargs)
+                unwrapped_tapes = [convert_to_numpy_parameters(t) for t in ctx.tapes]
```
Comment (albi3ro) [inline pennylane/interfaces/torch.py:190]: We should use tuples here instead.

**Review by albi3ro — COMMENTED** [review]
First point, I'm generally wondering if we should be using tuples instead of lists.  Not too important, but it moves us toward protecting this from being changed when they don't need to be changed.

Second, I'd prefer to rename the various `_copy_tape` helper methods to explain what they are actually doing, setting the parameters and unwrapping them to numpy.

**Review by albi3ro — APPROVED** [review]
Looks good 👍 

Love getting to remove a layer of nesting in the code too. Seems much more readable to me that way.

File: pennylane/interfaces/jax.py [hunk 4]
```diff
@@ -454,11 +457,10 @@ def execute_wrapper(params):
     @execute_wrapper.defjvp
     def execute_wrapper_jvp(primals, tangents):
         """Primals[0] are parameters as Jax tracers and tangents[0] is a list of tangent vectors as Jax tracers."""
-        new_tapes = [_copy_tape(t, a) for t, a in zip(tapes, primals[0])]
-
         if isinstance(gradient_fn, qml.gradients.gradient_transform):
+            at_max_diff = _n == max_diff
             _args = (
-                new_tapes,
+                set_parameters_on_copy_and_unwrap(tapes, primals[0], unwrap=at_max_diff),
```
Comment (rmoyard) [inline pennylane/interfaces/jax.py:463]: Minor: set it outside the function just for readability.

**Review by rmoyard — APPROVED** [review]

File: pennylane/interfaces/jax.py (hunk 4 above)
Comment (timmysilv) [reply pennylane/interfaces/jax.py]: moved outside of tuple

**Pull request merged**